draft-ietf-quic-retry-offload.md

```diff
@@ -174,43 +174,60 @@ the servers.
-{{?I-D.ietf-quic-load-balancers}}, it can verify that the Destination Connection
-ID is routable, and only admit non-Initial packets with routable DCIDs. As the
-Connection ID encoding is invariant across QUIC versions, the Retry Offload can
-do this for all short headers.
-
-Nothing in this section prevents Retry Offloads from making basic syntax
-correctness checks on packets with QUIC versions that it understands (e.g.,
-enforcing the Initial Packet datagram size minimum in version 1).
+## Retry Offloads with Per-Connection State
+
+A Retry Offload that keeps per-connection state can keep track of 5-tuples that
+it has accepted, either because it was in inactive mode or because it contained
+a valid token. It SHOULD accept subsequent Initial packets from these 4-tuples,
+regardless of the presence of a token, to avoid dropping part of the client's
+second flight.
+
+With per-connection state, the Retry Offload MAY drop Handshake and 0-RTT
+packets that do not correspond to an accepted 5-tuple. It SHOULD NOT drop short
+header packets, as these may be the result of a connection migration.
+
+## Retry Offloads without Per-Connection State
+
+Without per-connection state, Retry Offloads MUST admit all datagrams that begin
+with non-Initial packets.
+
+If a client Initial packet arrives without a Retry token, it might be the second
+client packet flight of a connection that was admitted when the Retry Offload
+was in inactive mode. If the Initial Packet does not consume the entire length
+of the UDP datagram, the Retry Offload MUST identify other packet types in the
+datagram. If there is a Handshake packet present, the Retry Offload MUST remove
+the Initial Packet and MAY remove any 0-RTT packet from the datagram, updating
+the UDP payload length accordingly. This prevents connection deadlock when
+client handshake packets progress the connection and the server is subject to
+the QUIC amplification limit.
+
+Nevertheless, there are two edge cases where Retry Offload can drop the second
+Initial packet and trigger a deadlock, when there is no per-connection state and
+the Retry Offload transitions from inactive to active mode mid-handshake.
+
+Note that {{RFC9000}} requires that clients ignore Retry packets after
+receiving a packet from the server other than Retry or Version Negotiation.
+
+1. The TLS Server Hello is spread over multiple Initial packets, and at least
+one of those is lost. Client Initial ACKs to recover the loss might be dropped
+by the Retry Offload, and the server cannot generate Handshake packets.
+
+2. The server sends a TLS Hello Retry Request. Subsequent communications will
+use Initial packets and potentially be dropped by the Retry Offload.
+
+Under the following conditions, the server SHOULD send CONNECTION_CLOSE instead
+of a Hello Retry Request or oversize Server Hello:
+
+* the first client Initial did not include a Retry Token;
+
+* the server cannot send additional packets due to the amplification limit; and
```

Comment:
I do not think we should do this. This is encouraging exceeding the amplification limit.

Reply:
A CONNECTION_CLOSE is small and IMO doesn't violate the spirit of the 3x limit, but I can instead add some language that the server has to save some space under the limit to send this.